smithy-model/src/main/java/software/amazon/smithy/model/shapes/Shape.java

```diff
@@ -88,10 +89,43 @@
         }
     }
 
-    protected void validateMixins(Collection<ShapeId> mixins) {
-        // This method is overridden in subclasses that allow mixins (structure, union, member).
-        // This situation can only happen by manipulating the semantic model in code.
-        throw new SourceException("Mixins are not allowed on " + getId() + ", a " + getType(), this);
+    protected void validateMixins(Map<ShapeId, Shape> mixins, Map<ShapeId, Trait> introducedTraits) {
+        Set<String> invalid = new TreeSet<>();
+        for (Shape mixin : mixins.values()) {
+            if (mixin.getType() != getType()) {
+                invalid.add(mixin.getId().toString());
+            }
+        }
+        if (!invalid.isEmpty()) {
+            String invalidList = String.join("`, `", invalid);
+            throw new IllegalStateException(String.format(
+                    "Mixins may only be mixed into shapes of the same type. The following mixins were applied to the "
+                            + "%s shape `%s` which are not %1$s shapes: [`%s`]", getType(), getId(), invalidList));
+        }
+    }
+
+    protected MemberShape getRequiredMixinMember(
+            AbstractShapeBuilder<?, ?> builder,
+            String name
+    ) {
+        MemberShape mixedMember = null;
+        for (Shape shape : builder.getMixins().values()) {
+            for (MemberShape member : shape.members()) {
+                if (member.getMemberName().equals(name)) {
+                    mixedMember = MemberShape.builder()
+                            .id(getId().withMember(name))
+                            .target(member.getTarget())
+                            .source(getSourceLocation())
+                            .addMixin(member)
+                            .build();
+                    break;
```

Comment:
just return here?

Reply:
No, we need the most recently introduced version. I added an explanatory comment.